Typescript migration

alcuadrado opened the pull request:
This is a WIP PR migrating this library to Typescript.

Things missing are:

1. Decide if we keep the json rpc functionality. Personally, as someone that works a lot with the RPC, I'd never expected that from this module.

2. Update code inline docs.

3. Keeping the callbacks-based API while modernizing the code (i.e. removing async) was kind of hard. Especially since browserify doesn't work well with `promisify` and `callbackify`. Maybe it's because of a combination of TS+borwserify+karma, but it made me lose a lot of time. I'm not sure if keeping that legacy API is worth the hassle. What do you think?

Some things to note:

* I defined a `Blockchain` interface instead of using `ethereumjs-blockchain` because this way this module can easily work with other implementations. Our blockchain implementation is too slow for many use-cases, so this is important. I'll submit a PR with something similar in the VM.

* I removed [this validation](https://github.com/ethereumjs/ethereumjs-block/blob/2c3908e5bbb6c9c94aeb5547a779295433f42545/index.js#L272), as it uses a deprecated method in `ethereumjs-blockchain`, which is actually a noop.

* I copied the common-handling logic from #69. Thanks, @micahriggan !

I'll probably be able to finish this tomorrow.

closes #71, #67, #63, #43, #69, #68

holgerd77 commented:
Great that your are tackling this so quickly, cool!

3. Definitely not worth keeping - we have e.g. already decided on the VM to completely drop the callback based API in exchange with @s1na and some extended community-expert-hearing 😛 - so just throw the old stuff away 😄, that would be a great along-improvement.

`Blockchain` interface: whew, that's a deep-dive-analytical change, pretty interesting to hear (`blockchain` slow) and very cool that you stumble and directly tackle stuff like this.

@alcuadrado @micahriggan Ah, I wrote [this](https://github.com/ethereumjs/ethereumjs-block/pull/69#issuecomment-515394932) comment before scanning over the work and explanations here, so comment seems to be already outdated respectively answered through the active process here.

@micahriggan Thanks for your work on the original PR and working out the problem space more clearly together!

holgerd77 commented on .travis.yml:
```diff
@@ -28,9 +29,19 @@ matrix:
+      node_js: "10"
+      env: CXX=g++-4.8 TEST_SUITE=test:node
+    - os: linux
+    node_js: "11"
```
Just a note in case you didn't notice yet: Travis couldn't complete - eventually some error along with the config file updates. Maybe it's this indentation here?

alcuadrado replied:
This was pretty hard to debug. Thanks for noticing it :)

alcuadrado commented:
> 1. Definitely not worth keeping - we have e.g. already decided on the VM to completely drop the callback based API in exchange with @s1na and some extended community-expert-hearing 😛 - so just throw the old stuff away 😄, that would be a great along-improvement.

Awesome! 🎉

> I wrote [this](https://github.com/ethereumjs/ethereumjs-block/pull/69#issuecomment-515394932) comment before scanning over the work and explanations here

Oops, same here. I answered in the other PR.

> `Blockchain` interface: whew, that's a deep-dive-analytical change, pretty interesting to hear (`blockchain` slow) and very cool that you stumble and directly tackle stuff like this.

I'll open an issue with an explanation in the blockchain repo later today.

alcuadrado commented:
I think this PR is ready to be reviewed now.

I ported the RPC-related functions, removed the callback-based API, fix the travis config, and documented the missing uncles' validation.

As a summary that could be added to the changelog, the main changes that this PR introduces are:

1. The codebase has been ported to TypeScript.
2. This library only offers a `Promise`-based API now.
3. Errors now are instances of `Error` instead of strings.
3. Updated `ethereum-util` and `ethereumjs-tx` to their latest versions.

alcuadrado commented:
Please, take a look at the initial message from this PR, as it's relevant here: https://github.com/ethereumjs/ethereumjs-vm/pull/566

holgerd77 commented on src/header.ts:
```diff
@@ -0,0 +1,359 @@
+      })
+    })
+  }
+}
```
Ok, looks good, promisification of the `validate` function.

holgerd77 commented on src/block.ts:
```diff
@@ -0,0 +1,269 @@
+      // a [Buffer[], Buffer[], Buffer[]]
+      const dataAsAny = rlp.decode(data) as any
+      data = dataAsAny as [Buffer[], Buffer[], Buffer[]]
+    }
```
I will just believe this. 😇

alcuadrado replied:
Actually, none of the libraries validate their input. They just break in unexpected ways if these assumptions don't hold.

holgerd77 commented on src/block.ts:
```diff
@@ -0,0 +1,269 @@
+
+    if (Array.isArray(data)) {
+      // TODO: Pass the common object
+      this.header = new BlockHeader(data[0], opts)
```
Is this difficult to directly pass the common object here instead of adding a placeholder TODO statement?

holgerd77 replied:
(and instances below)

alcuadrado replied:
It's not a problem here, but when initializing the transactions. The block defaults its hardfork to `null`, and `Transaction` doesn't support it.

I will pass the common object to the headers, and not to the transaction.

alcuadrado replied:
I also two TODOs explaining the problems around Common, Transaction and block numbers

holgerd77 commented on src/block.ts:
```diff
@@ -0,0 +1,269 @@
+   * Produces a serialization of the block.
+   *
+   * @param rlpEncode - If `true`, the returned object is the RLP encoded data as seen by the
+   * Ethereum wire protocol. If `true`, a tuple with the raw data of the header, the txs and the
```
Second `true` in the comment seems to be wrong if I read correctly.

alcuadrado replied:
The second `true` was false. Thanks.

holgerd77 commented on src/block.ts:
```diff
@@ -0,0 +1,269 @@
+   *
+   * @param blockChain - the blockchain that this block wants to be part of
+   */
+  async validateUncles(blockchain: Blockchain) {
```
Return value is missing here.

alcuadrado replied:
Adding it made me realize that the return type was super bizarre. I simplified this function.

holgerd77 commented on src/block.ts:
```diff
@@ -0,0 +1,269 @@
+
+    const height = new BN(this.header.number)
+    return uncleHeader.validate(blockchain, height)
+  }
```
I am just not getting the whole picture why this "uncle already included" could be removed, could you give a really quick reasoning?

alcuadrado replied:
It shouldn't be removed, but there's no way to validate it using `ethereumjs-blockchain`. The previous code was just calling a function of the blockchain that was a no-op. 

This really surprised me, so I spent some time trying to figure out how it happened, but couldn't found the reason. All I know is that that functionality was removed in this PR: https://github.com/ethereumjs/ethereumjs-blockchain/pull/47/files#diff-168726dbe96b3ce427e7fedce31bb0bcR481

holgerd77 replied:
//cc @vpulim

holgerd77 commented on src/block.ts:
```diff
@@ -0,0 +1,269 @@
+    const height = new BN(this.header.number)
+    return uncleHeader.validate(blockchain, height)
+  }
+}
```
Rest/generally looks good, really nice improvements all over the board.

holgerd77 commented on test/header.ts:
```diff
@@ -0,0 +1,104 @@
+    )
+    st.end()
+  })
+})
```
Test files ok, some updates along the new promisified API.

alcuadrado commented on .travis.yml:
```diff
@@ -1,20 +1,21 @@
-  - CXX=g++-4.8
+  - "10"
+  - "11"
+  - "12"
```
I just noticed that 11 is not supported anymore. I'll remove it.

alcuadrado commented on karma.conf.js:
```diff
@@ -1,49 +1,18 @@
+      },
     }
   })
 }
```
This may be my fault, I copied part of the config from the `-tx` migration.

alcuadrado replied:
I think they were disabled because travis was failing because the tests' output exceeded the maximum log that they are willing to store. I tried different things and finally made it work by changing the Karma reporters setting.

alcuadrado commented on package.json:
```diff
@@ -1,19 +1,29 @@
 {
   "name": "ethereumjs-block",
-  "version": "2.2.0",
+  "version": "3.0.0",
```
I reverted this change

alcuadrado commented on package.json:
```diff
@@ -35,28 +45,36 @@
-    "karma-tap": "^4.0.0",
-    "standard": "^8.4.0",
-    "tape": "^4.2.0"
+    "karma-tap": "^4.1.4",
```
Karma now loads plugins automatically if they are installed in the same `node_modules` and their name starts with `karma-`.

alcuadrado commented on src/header.ts:
```diff
@@ -0,0 +1,359 @@
+import { BN } from 'ethereumjs-util'
+import { Blockchain, BlockHeaderData, BufferLike, NetworkOptions, PrefixedHexString } from './types'
+import { Buffer } from 'buffer'
+import { Block } from './block'
```
This also seems weird to me, as it forces a cyclic dependency between `block.ts` and `header.ts`.  

We could change the methods to operate on `Header`s, but the blockchain module also operates on `Block`s, so that cyclic dependency would still be needed.

alcuadrado commented on .travis.yml:
```diff
@@ -1,20 +1,21 @@
-    - g++-4.8
+      - g++-4.8
 before_install:
   - sh -e /etc/init.d/xvfb start
```
Done

alcuadrado commented:
@holgerd77 Ia addressed all the issues. Just not sure if something different can be done about [this one](https://github.com/ethereumjs/ethereumjs-block/pull/72#discussion_r309384919).

alcuadrado commented on src/from-rpc.ts:
```diff
@@ -0,0 +1,66 @@
+ *
+ * @param blockParams - Ethereum JSON RPC of block (eth_getBlockByNumber)
+ * @param uncles - Optional list of Ethereum JSON RPC of uncles (eth_getUncleByBlockHashAndIndex)
+ * @param networkOptions - An object describing the network
```
Oops, I missed this comment. You are right, it can be really confusing. I'll rename it.

micahriggan commented:
Amazing 😍
